lib/validators.rb

```diff
@@ -137,9 +137,107 @@ def check_shipments_specificities
       }
     end
 
+    def vehicle_days_availability(vehicles_set, timewindows_set)
+      vehicles_days = timewindows_set.collect{ |timewindow|
+        if timewindow.empty?
+          []
+        else
+          days = timewindow.flat_map{ |tw| tw[:day_index] }
+          days.compact!
+          days.uniq
+        end
+      }.delete_if(&:empty?)
+
+      vehicles_unavailable_indices = vehicles_set.collect{ |v| v[:unavailable_work_day_indices] }
+      vehicles_unavailable_indices.compact!
+
+      [vehicles_days, vehicles_unavailable_indices]
+    end
+
+    def check_vehicle_trips_stores_consistency(relation_vehicles)
+      relation_vehicles.each_with_index{ |vehicle_trip, v_i|
+        case v_i
+        when 0
+          unless vehicle_trip[:end_point_id]
+            raise OptimizerWrapper::DiscordantProblemError.new('First trip should at least have end point id')
+          end
+        when relation_vehicles.size - 1
+          unless vehicle_trip[:start_point_id]
+            raise OptimizerWrapper::DiscordantProblemError.new('Last trip should at least have start point id')
+          end
+        else
+          unless vehicle_trip[:start_point_id] && vehicle_trip[:end_point_id]
+            raise OptimizerWrapper::DiscordantProblemError.new('Trip should have start and end point id')
+          end
+        end
+
+        next if v_i.zero? ||
+                vehicle_trip[:start_point_id] == relation_vehicles[v_i - 1][:end_point_id] ||
+                (@hash[:points].find{ |pt| pt[:id] == vehicle_trip[:start_point_id] }[:location] ==
+                  @hash[:points].find{ |pt| pt[:id] == vehicle_trip[:end_point_id] }[:location])
+
+        raise OptimizerWrapper::DiscordantProblemError.new('One trip should start where previous trip ended')
+      }
+    end
+
+    def check_vehicle_trips_timewindows_consistency(relation_vehicles)
+      vehicles_timewindows =
+        relation_vehicles.collect{ |v| v[:timewindow] ? [v[:timewindow]] : v[:sequence_timewindows].to_a }
+
+      return if vehicles_timewindows.all?(&:empty?)
+
+      week_days, unavailable_indices = vehicle_days_availability(relation_vehicles, vehicles_timewindows)
+      if week_days.uniq.size > 1 ||
+         unavailable_indices.uniq.size > 1
+        raise OptimizerWrapper::DiscordantProblemError.new(
+          'Vehicles in vehicle_trips relation have different available days'
+        )
+      end
+
+      vehicles_timewindows.each_with_index{ |v_tw, v_i|
+        next if v_i.zero?
+
+        v_tw.each{ |tw|
+          day = tw[:day_index]
+          previous_tw =
+            vehicles_timewindows[v_i - 1].select{ |ptw|
+              ptw[:day_index].nil? ||
+                ptw[:day_index] == day
+            }.min_by{ |ptw| ptw[:start] }
```

Comment:
Just to verify: this ``select.min_by` can't return `nil` because previous_vehicle should have the same timewindows with the "current" vehicle (`v_tw`), is that it? otherwise, if the select returns empty, min_by returns `nil`.

Reply:
Here, min_by can return _nil_. So you are right, my next condition needs to be fixed : we need to ensure there is one previous_tw.